test/test/boot.js

```diff
@@ -404,6 +404,119 @@ function configureJasmineEnvironment() {
+ * via this shim.
+ */
+function patchCastCanDisplayType() {
+  if (!navigator.userAgent.includes('CrKey')) {
+    // Not Chromecast.  Do nothing.
+    return;
+  }
+
+  // Create the namespaces if needed.
+  if (!window.cast) {
+    window.cast = {};
+  }
+  if (!cast.__platform__) {
+    cast.__platform__ = {};
+  }
+
+  if (cast.__platform__.canDisplayType) {
+    // Already exists, do nothing.
+    return;
+  }
+
+  // Create an async shim.  Calls to canDisplayType will be translated into
+  // async messages to the top frame, which will then execute the method and
+  // post a message back with results (or an error).  The resolve/reject
+  // functions for the shim's returned Promise will be stored temporarily in
+  // these maps and matched up by request ID.
+  /** @type {!Map<number, function()>} */
+  const resolveMap = new Map();
+  /** @type {!Map<number, function()>} */
+  const rejectMap = new Map();
+  /** @type {number} */
+  let nextId = 0;
+
+  // Listen for message events for results/errors from the top frame.
+  window.addEventListener('message', (event) => {
+    const data = event.data;
+    console.log('Received cross-frame message', data);
+
+    if (data.type == 'cast.__platform__:result') {
+      // Find the matching resolve function and resolve the promise for this
+      // request.
+      const resolve = resolveMap.get(data.id);
+      if (resolve) {
+        resolve(data.result);
+
+        // Clear both resolve and reject from the maps for this ID.
+        resolveMap.delete(data.id);
+        rejectMap.delete(data.id);
+      }
+    } else if (data.type == 'cast.__platform__:error') {
+      // Find the matching reject function and reject the promise for this
+      // request.
+      const reject = rejectMap.get(data.id);
+      if (reject) {
+        reject(data.result);
+
+        // Clear both resolve and reject from the maps for this ID.
+        resolveMap.delete(data.id);
+        rejectMap.delete(data.id);
+      }
+    }
+  });
+
+  // Shim canDisplayType to proxy the request up to the top frame.
+  cast.__platform__.canDisplayType = function() {
+    return new Promise((resolve, reject) => {
+      // Craft a message for the top frame to execute this method for us.
+      const message = {
+        id: nextId++,
+        type: 'cast.__platform__',
+        command: 'canDisplayType',
+        args: Array.from(arguments),
+      };
+
+      // Store the resolve and reject functions so we can act on results/errors
+      // later.
+      resolveMap.set(message.id, resolve);
+      rejectMap.set(message.id, reject);
+
+      // Reject after a 5s timeout.  This can happen if we're running under an
+      // incompatible version of Chromecast WebDriver Server's receiver app.
+      setTimeout(() => {
+        reject(new Error('canDisplayType timeout!'));
+
+        // Clear both resolve and reject from the maps for this ID.
+        resolveMap.delete(message.id);
```

Comment:
nit; I guess you can just create one `Promise` map.

Reply:
No, you can't.  Promises don't have accessors for the resolve and reject functions passed to the callbacks in their constructors.  You're thinking of `goog.promise.Resolver`, which we don't have here.